From 82483b08953d6aab2af233f3e753b524cad8985b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fabiano=20Fid=C3=AAncio?= <fidencio@redhat.com>
Date: Wed, 22 Jul 2015 03:57:18 +0200
Subject: [PATCH] events: don't create glib IO watch for disabled handles

It's possible to create a handle to watch for file events which do not
watch for any file event. Such a handle can be enabled later with
virt_viewer_events_update_handle() by setting some conditions to watch for.

When a handle is disabled after it has been created,
virt_viewer_events_update_handle() makes sure it removes the corresponding
virt_viewer_events_handle::source IO watch if any was set.
virt_viewer_events_add_handle() will always create a
virt_viewer_events_handle::source IO watch even if the handle is not
watching for any events.

This commit makes consistent by only creating a watch with g_io_add_watch()
when the caller asked to watch for some events.

Based on commit d71c143936a35cd6c3f23ae0cbf7f3215d944051 from
libvirt-glib.
Original author: Christophe Fergeau <cfergeau@redhat.com>

Related to: rhbz#1243228

(cherry picked from commit 3281c47e8b03cd0072c5ac18e97f1a36d08ec485)
---
 src/virt-viewer-events.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/src/virt-viewer-events.c b/src/virt-viewer-events.c
index f68791f..fe3aaa6 100644
--- a/src/virt-viewer-events.c
+++ b/src/virt-viewer-events.c
@@ -113,10 +113,12 @@ int virt_viewer_events_add_handle(int fd,
 
     g_debug("Add handle %d %d %p", data->fd, events, data->opaque);
 
-    data->source = g_io_add_watch(data->channel,
-                                  cond,
-                                  virt_viewer_events_dispatch_handle,
-                                  data);
+    if (events != 0) {
+        data->source = g_io_add_watch(data->channel,
+                                      cond,
+                                      virt_viewer_events_dispatch_handle,
+                                      data);
+    }
 
     g_ptr_array_add(handles, data);
 
